### PR TITLE
[SPARK-30927][SS]StreamingQueryManager should avoid keeping reference to terminated StreamingQuery

### PR DESCRIPTION
**Pull request by zsxwing**
### What changes were proposed in this pull request?

Right now `StreamingQueryManager` will keep the last terminated query until `resetTerminated` is called. When the last terminated query has lots of states (a large sql plan, cached RDDs, etc.), it will keep a lot of memory unnecessarily. Actually, what `StreamingQueryManager` really needs is just the exception of the last failed query.

This PR changes the internal field `lastTerminatedQuery` in `StreamingQueryManager` to remember the last exception rather than the query to save the memory.

### Why are the changes needed?

Avoid keeping memory unnecessarily.

### Does this PR introduce any user-facing change?

No

### How was this patch tested?

This PR doesn't change any public behaviors. The existing tests have covered the touched codes.

### CODE REVIEW

**Comment by zsxwing** [comment]
cc @brkyvz

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryManager.scala [hunk 1]
```diff
@@ -125,11 +134,11 @@ class StreamingQueryManager private[sql] (sparkSession: SparkSession) extends Lo
   @throws[StreamingQueryException]
   def awaitAnyTermination(): Unit = {
     awaitTerminationLock.synchronized {
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryManager.scala]: While we're here, how about simply calling `awaitAnyTermination(Long.MAX_VALUE)` which is effectively same?

**Review by HeartSaVioR — APPROVED** [review]
LGTM, only optional minor.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.0.

**Pull request closed without merging**